cmd/influx_inspect/buildtsi/buildtsi.go

```diff
@@ -152,25 +208,50 @@ func (cmd *Command) processRetentionPolicy(sfile *tsdb.SeriesFile, dbName, rpNam
 			continue
 		}
 
-		if err := cmd.processShard(sfile, dbName, rpName, shardID, filepath.Join(dataDir, fi.Name()), filepath.Join(walDir, fi.Name())); err != nil {
+		shards = append(shards, struct {
+			ID   uint64
+			Path string
+		}{shardID, fi.Name()})
+	}
+
+	errC := make(chan error, len(shards))
+	var maxi uint32 // index of maximum shard being worked on.
+	for k := 0; k < cmd.concurrency; k++ {
+		go func() {
+			for {
+				i := int(atomic.AddUint32(&maxi, 1) - 1) // Get next partition to work on.
+				if i >= len(shards) {
+					return // No more work.
+				}
+
+				id, name := shards[i].ID, shards[i].Path
+				errC <- cmd.processShard(sfile, dbName, rpName, id, filepath.Join(dataDir, name), filepath.Join(walDir, name))
+			}
+		}()
+	}
```

Comment:
Is this something we can generalize? It seems like a common approach we do all over the place.

Reply:
I think so yeah, but should probably do it in a future PR and apply it to the entire codebase.